### PR TITLE
Migrate to main

### LINKED ISSUE #145: Transition from master to main
Switch from using master to using main instead.

### PR DESCRIPTION
**Pull request by TroyWalshProf**
Closes #145

Move full CI to Actions.
Use one pipeline for CI and CD
Add simple env approval for releasing Selenium.Axe to nuget.org

### CODE REVIEW

**Comment by dbjorge** [comment]
@TroyWalshProf , I'm assuming this is still in progress based on the commit history and the failing CI build; feel free to give @lisli1 and I an @ mention once this is ready to review and we'd be happy to take a look!

**Comment by TroyWalshProf** [comment]
@dbjorge and @lisli1 - Finally got this all working and ready for review.

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:140]: test-reporter's docs suggests that this is probably going to have problems for folks sending PRs from forks of the repo, since the implicit github token used there won't have rights to create checks. [They document](https://github.com/dorny/test-reporter#recommended-setup-for-public-repositories) a workaround that involves moving this this "report" job out to a different trigger

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:10]: The name `override` confused me a bit when I read it later in the file. Maybe `version` or `version_override` would be more clear?

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:50]: It would probably be good to make sure the step will fail if this script fails, rather than continuing with a corrupted version number like I suspect it will right now

Comment (dbjorge) [reply .github/workflows/mainPipeline.yml:50]: Ah, nevermind this comment; [it appears](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsrun) GitHub Actions prepends `$ErrorActionPreference = 'Stop'` automatically.

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:99]: The results report only distinguishes the 3 target frameworks' worth of tests by their filenames. The [report documentation](https://github.com/dorny/test-reporter#supported-formats) suggests that, rather than using a `--results-directory` argument here, it might be preferable to set the result directory and/or result filename using `.csproj` properties so that `$(TargetFramework)` can be embedded into the file path; that would make it much easier to tell in the report which set of results is which. The docs suggest that the `.csproj` syntax is something like this (I haven't tested it):

```xml
<PropertyGroup>
  <VSTestLogger>trx%3bLogFileName=$(MSBuildProjectName).trx</VSTestLogger>
  <VSTestResultsDirectory>$(MSBuildThisFileDirectory)/TestResults/$(TargetFramework)</VSTestResultsDirectory>
</PropertyGroup>
```

**Review by dbjorge — CHANGES_REQUESTED** [review]
Looks great overall! Left a few suggestions inline

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml]: Moved to separate action

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml]: Updated with framework specific trx file names

File: .github/workflows/test-report.yml
Comment (dbjorge) [inline .github/workflows/test-report.yml:5]: `MigrateToMainn` probably isn't necessary here (if you do want to include it, there's a typo)

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:131]: I worry that doing this on every push to main makes it easy to accidentally issue patch releases (the release-drafter default) for things that aren't really patch-level changes. Maybe this should be paired with some sort of check that forbids merging PRs that haven't been explicitly labelled as major/minor/patch, and/or some mechanism for non-user-facing changes that shouldn't trigger releases?

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:130]: It doesn't seem like using the `buildNumber` variable in the job name is cooperating; I'm seeing the placeholder string in the check name.

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:54]: Maybe a good idea to also `Write-Host` the version without the special `::set-output` syntax, so it gets logged in a more obvious spot than "buried in the Build and Analyze step's command" for debugging purposes?

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [inline .github/workflows/mainPipeline.yml:88]: This is going to lock to a potentially-outdated version of the scanner basically indefinitely. Perhaps the key here should include a version number for the scanner tool?

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml:130]: It does corporates if that job isn't skipped

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml:131]: This is why we have a protection rule on the nuget environment.
This way every PR is treated as a release candidate and it is up to the reviews if they want to allow the push to nuget.org or not.

Somewhat related, how would you feel about being one of the approvers?

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml:88]: Not sure how we would get stuck on an outdated version, we are aren't hosting our own agents.

File: .github/workflows/test-report.yml
Comment (TroyWalshProf) [reply .github/workflows/test-report.yml]: You are correct

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml:55]: Added

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml:132]: Here is how it will look once me get this merged into main:
image
*From a different project that does the same thing

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [reply .github/workflows/mainPipeline.yml:89]: Cached directories can be shared between different builds, even using GitHub-provided agents. Because the `Install SonarCloud scanner` step uses `if: steps.cache-sonar-scanner.outputs.cache-hit != 'true'`, it's going to be skipped if there exists a previously-cached `.\.sonar\scanner` directory with key `${{ runner.os }}-sonar-scanner`. That means that the first time a build runs, it'll cache the scanner folder with key `Windows-sonar-scanner`, and since that key name isn't based on the scanner version, whatever version was cached will keep being re-used in subsequent builds (until the cache entry happens to be evicted).

If you encode the scanner version number into the `key`, you'll only get a cache hit if the cached version matches the installed version. That's the behavior you want in combination with the later `if: steps.cache-sonar-scanner.outputs.cache-hit != 'true'`, since it'll install the updated version and update the cache when a different version number is present.

`restore-keys` is intended for use with partial matches; I don't think it's necessary to use it at all here if it's exactly equal to the `key.

If the scanner impl and its cache are going to be cached at the same cadence, you can also avoid some of the repetition by using `actions/cache@v2` instead of `@v1`; v2 supports passing multiple paths in the same step.

[GitHub's docs](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows) are a good overview to learn more.

File: .github/workflows/mainPipeline.yml
Comment (dbjorge) [reply .github/workflows/mainPipeline.yml:132]: Ah, got it, I didn't realize that setting the environment property would have that effect. Neat!

Comment (dbjorge) [reply .github/workflows/mainPipeline.yml:132]: > Somewhat related, how would you feel about being one of the approvers?

This is fine by me!

File: .github/workflows/mainPipeline.yml
Comment (TroyWalshProf) [reply .github/workflows/mainPipeline.yml]: Oh, I finally  get what you are saying.
Just pulled the caching as it saves almost no time.

**Review by dbjorge — APPROVED** [review]

**Pull request merged**
